server/server.go

```diff
@@ -235,15 +235,16 @@ func (a *ArgoCDServer) Run(ctx context.Context, port int, metricsPort int) {
 	var httpS *http.Server
 	var httpsS *http.Server
 	if a.useTLS() {
-		httpS = newRedirectServer(port)
+		httpS = newRedirectServer(port, a.RootPath)
 		httpsS = a.newHTTPServer(ctx, port, grpcWebS)
 	} else {
 		httpS = a.newHTTPServer(ctx, port, grpcWebS)
 	}
 	if a.RootPath != "" {
-		httpS.Handler = withRootPath(httpS.Handler, a.RootPath)
+		httpS.Handler = withRootPath(httpS.Handler, a)
+
 		if httpsS != nil {
-			httpsS.Handler = withRootPath(httpsS.Handler, a.RootPath)
+			httpsS.Handler = withRootPath(httpsS.Handler, a)
```

Comment:
As I understand now we have two set of handlers on both `/` and `/<rootPath>` (e.g `/healthz` and `/argocd/healthz`, `/api/webhook` and `/argocd/api/webhook` etc). I think this is too much. I suggest to duplicate only `/healthz`. For example you can register health url using `healthz.ServeHealthCheck` within using `withRootPath` method:

```golang
func (a *ArgoCDServer) withRootPath(handler http.Handler, root string) http.Handler {
	// get rid of slashes
	root = strings.TrimRight(strings.TrimLeft(root, "/"), "/")

	mux := http.NewServeMux()
	mux.Handle("/"+root+"/", http.StripPrefix("/"+root, handler))
	healthz.ServeHealthCheck(mux, func() error {
		_, err := a.KubeClientset.(*kubernetes.Clientset).ServerVersion()
		return err
	})
	return mux
}	
```